### ISSUE TITLE
New feature: ng-switch-when-separator in Angular 2 to allow multiple values

### ISSUE THREAD

**Issue by peterhe2000** [angular/angular#14659]
Also mentioned in here:
https://github.com/angular/angular/issues/11926

Angular version: 2.4.7

**Comment by kanlidy** [comment]
@peterhe2000 I support you.

I am building an angular application like the below.


```
<div [ngSwitch]="data.type">
  <div *ngSwitchCase="'multi-choice'">FORM 1</div>
  <div *ngSwitchCase="'singe-choice'">FORM 1</div>
  <div *ngSwitchCase="'range'">FORM 2</div>
</div>
```


In angular1, I can do this:

`<div class="animate-switch" ng-switch-when="'multi-choice'|'singe-choice'" ng-switch-when-separator="|">Settings Div</div>`

But, there's no "ng-switch-when-seprartor" feature like in angular2

```
<div [ngSwitch]="data.type">
  <div *ngSwitchCase="'multi-choice' || 'singe-choice'">FORM 1</div>
</div>
```

for now, the solution is:

```
<div [ngSwitch]="data.type">
    <div *ngSwitchCase="data.type=='multi-choice' ||data.type== 'singe-choice'?data.type:'notInCase'">FORM 1</div>
<div *ngSwitchCase="'range'">FORM 2</div>
</div>
```

**Comment by peterhe2000** [comment]
@kanlidy You solution may not work according to https://github.com/angular/angular/issues/12174.

**Comment by waterplea** [comment]
I'd also love that feature, since plain switch case in every language provide that functionality.

**Comment by rockmacaca** [comment]
Hi, I'm also interested in this! I also just had the thought of borrowing the new `NgIf/Then/Else` microsyntax here since it kind of makes sense for `NgSwitch` to be able to use a `TemplateRef`. This issue could be solved by allowing the following syntax:

```html
<div [ngSwitch]="data.type">
  <div *ngSwitchCase="'multi-choice'; do form1"></div>
  <div *ngSwitchCase="'singe-choice'; do form1"></div>
  <ng-template #form1>FORM 1</ng-template>
  <div *ngSwitchCase="'range'">FORM 2</div>
</div>
```

Additionally, it allows for some other neat things that you can do with `TemplateRef`s.
Obviously it's not quite as terse as most languages' `switch` keyword but it seems to fit well. Also I can't think of a better keyword than `do` at the moment.

**Comment by Sky4CE** [comment]
any progress on this ?

**Comment by Ghostbird** [comment]
I'd like to see this too, I was surprised that angular does not offer this functionality, since it is one of the basic functionalities of switch statements in most languages.
For now I'll rewrite my code with *ngIf statements.

**Comment by Ghostbird** [comment]
Would it be sufficient to replace:
```typescript
 _matchCase(value: any): boolean {
const matched = value == this._ngSwitch;
```
with:
```typescript
 _matchCase(value: any): boolean {
const matched = value == this._ngSwitch || (value instanceof Array && value.some(x => x == value));
```
in [packages/common/src/directives/ng_switch.ts](../blob/master/packages/common/src/directives/ng_switch.ts)

I guess it isn't that simple.

In the meantime I'm going to try the boolean evaluation trick mentioned by @kanlidy, thanks for that.

**Comment by rodgerpaulo** [comment]
+1 on this

**Comment by pmccloghrylaing** [comment]
I don't mind the `NgIf` (`...; do ...`) syntax, that seems like a reasonable solution.

Anything else requires a separator, the best I could think of was `or`. The boolean and bitwise `OR` operators (`||` and `|`) should behave as per usual in JS. Or you require the separator to be specified like in v1.

Another option could be defining the cases in child elements, e.g.:
```html
<div [ngSwitch]="data.type">
  <div *ngSwitchMultiple>
    <ng-switch-case value="'multi-choice'"></ng-switch-case>
    <ng-switch-case value="'single-choice'"></ng-switch-case>
    FORM 1
  </div>
  <div *ngSwitchCase="'range'">FORM 2</div>
</div>
```
I think in comparison using the `NgIf` `do` is far simpler.

**Comment by trotyl** [comment]
Relates to #20027 (which is possible while this one is not)

**Comment by specialkk** [comment]
Yeah I don't think the separator approach is that great of an idea. #20027 also seems reasonable.

Also, if anyone following this wants an actual workaround they can do now without having to duplicate code, [`NgTemplateOutlet`](https://angular.io/api/common/NgTemplateOutlet) can be used in the style I suggested before, though it's still a lot of boilerplate:

```html
<div [ngSwitch]="data.type">
  <div *ngSwitchCase="'multi-choice'">
    <ng-container *ngTemplateOutlet="form1"></ng-container>
  </div>
  <div *ngSwitchCase="'single-choice'">
    <ng-container *ngTemplateOutlet="form1"></ng-container>
  </div>
  <ng-template #form1>FORM 1</ng-template>

  <div *ngSwitchCase="'range'">FORM 2</div>
</div>
```

**Comment by kkajdd** [comment]
HI,
When i try to implement ngSwitch it works like charm but in the spec file i get an error as No provider for NgSwitch? what might be the cause?

**Comment by weskenyon** [comment]
@specialkk For now I think this is a good approach. Its not ideal but it works and is what I ended up doing until I just fixed our switch statements to not need this. I would prefer somehow if we could move that template HTML into another file somehow. It feels odd just having your template at the bottom of your html (or wherever) to reference.

**Comment by santialbo** [comment]
In order to keep it backwards compatible. Would it be acceptable to hae something like the following:

```html
<div [ngSwitch]="data.type">
  <div *ngSwitchCases="['multi-choice', 'single-choice']">FORM 1</div>
  <div *ngSwitchCase="'range'">FORM 2</div>
</div>
```
Notice it's a new directive ngSwitchCase*s*.

**Comment by pmccloghrylaing** [comment]
@santialbo I like the sound of that. Implementation should be quite simple as well.

**Comment by trotyl** [comment]
That's https://github.com/angular/angular/issues/20027 then.

**Comment by adamdport** [comment]
@santialbo IMHO it doesn't need to be a new directive. Why not leave it ngSwitchCase? If the case is an array, check each value, otherwise use the existing functionality

**Comment by pkozlowski-opensource** [comment]
> IMHO it doesn't need to be a new directive. Why not leave it ngSwitchCase? If the case is an array, check each value, otherwise use the existing functionality

What if you would like to match on an array _instance_ ? Doing what you propose would break this use-case. So this needs to be a new directive to clearly indicate that we want to match on multiple values.

**Comment by jonrimmer** [comment]
Until this is merged, I created a standalone directive that you can add to your project:

https://gist.github.com/jonrimmer/eaabd619e2edeaebed83b7bc68f33daf

Use it like this:

```html
<div [ngSwitch]="value">
  <div *jrSwitchCases="['foo', 'bar']">
    Foo or bar
  </div>
</div>
```

**Comment by maxigimenez** [comment]
https://github.com/angular/angular/pull/27421 add the support of array on `*ngSwithCase` and is fully compatible with the current version.

**Comment by IlCallo** [comment]
Any progress in this?
Considering that we are now going to Angular 8, this could be implemented even if it's a breaking change.
And if not Angular 8 (it's a bit late), we can put it inside Angular 9.

**Comment by Lakston** [comment]
Apparently it's not in A9 either, I'd love to see this feature someday I just had to hack it by doing this:

```

**Comment by samclaus** [comment]
Just to offer some framework cross pollination:

Svelte does not currently have a `{#switch}` directive. However, it does have full-fledged `if-else-if-else-if-...-else` logic support in templates. By comparison, Angular's over-engineered `*ngIf` "microsyntax" with support for `else` via a `TemplateRef` is a real PITA. I have not once needed to write a custom structural directive and I don't ever plan to: keeping any sort of complicated logic in TypeScript and keeping templates as a shallow/dumb layer which only handles matching on basic view state possibilities avoids uncanny-valley reimplementations of too many JS features in the Angular template compiler (why do the same work twice when the languages can work in tandem?).

The only reason I find myself on this old `ngSwitch` issue is because Angular's barebones `if-else` template logic is not very capable and requires you to write several `<ng-templates>` to achieve the logic chaining, although you _can_ flatten template elements to the outermost level of Angular-HTML to avoid indentation hell so it is not the end of the world at least.

I think making `*ngIf` more powerful would alleviate many problems. [Svelte is waiting on JS `match` syntax to be formalized before implementing a `switch`/`match` feature in templates](https://github.com/sveltejs/rfcs/pull/49), but it **doesn't matter** because their `if` functionality closes the gap and effectively makes such a feature nothing more than a cherry on top. And if you do not understand why `switch` is a bad--of course, it was made with good intentions--general language construct and `match` is a natural evolution for it, please see [my comment](https://github.com/sveltejs/rfcs/pull/49#issuecomment-875880779) on the aforementioned issue.

In any case, Angular currently has no way to nicely write a view with many different but flat (non-nested) states where some of the states overlap and want to re-use the same template fragment (chunk of code). It's an older framework and I want to be clear I'm NOT just here to talk s**t; this is intended as constructive criticism and I think Angular did a good job at a lot of things but some inevitable missteps were made. I don't know how the Angular team feels about large-scale template language changes vs. 80% fixes via the "standard library" but I do recall from [other issues](https://github.com/angular/angular/issues/39525#issuecomment-842669317) I have written that there is a somewhat-reasonable fear of introducing non-HTML-compliant syntax to templates because that has a good chance of breaking tooling which parses the templates as HTML. Of course, such tooling could be updated to use an Angular-provided parser which is a superset of HTML. Food for thought.

**Comment by samclaus** [comment]
@petebacondarwin That's awesome! Sounds like you guys have a good game plan. Of course it will take time. I don't think any serious developers would shame a team for taking a while to implement things cleanly. 😃

**Comment by HyperLife1119** [comment]
The new control flow syntax still doesn't seem to support this feature.

**Comment by Harpush** [comment]
Ideally something like:
```html
@case (1)
@case (2) {
  ...
}
```

**Comment by twittwer** [comment]
Or a list of options
```
@case (1, 2) {
  ...
}
```

**Comment by ilyakonrad** [comment]
Surprised to not see this supported by the new control flow syntax, it's been awaited for over 7 years now 😞

**Comment by nook24** [comment]
Can't believe that this is **still** not possible

**Comment by ko1ebayev** [comment]
seems like it works with enums

**Comment by BastienVH** [comment]
> seems like it works with enums

Can you please give an example of the syntax that works with enums?
I've looked around and tried some things, but can't get it to work.
What version of Angular are you on?

**Comment by Para-FR** [comment]
![Adam Lambert So What](https://github.com/user-attachments/assets/abd9a073-2f6b-4877-8af2-00c22073a7ef)

```@for (column of props().columns; track $index) {

            @switch (column.field) {
              @case ('status') {
                <td>
                  <app-badge [props]="{ label: item[column.field] }"/>
                </td>
              }
              @case ('date' || 'createdAt' || 'updatedAt') {
                <td>{{ item[column.field] | date }}</td>
              }
              @default {
                <td>{{ item[column.field] }}</td>
              }
            }
          }```

**Comment by sklei** [comment]
> @for (column of props().columns; track $index) {
> 
>             @switch (column.field) {
>               @case ('status') {
>                 <td>
>                   <app-badge [props]="{ label: item[column.field] }"/>
>                 </td>
>               }
>               @case ('date' || 'createdAt' || 'updatedAt') { // or @case 1 @case 2 @case 3
>                 <td>{{ item[column.field] | date }}</td>
>               }
>               @default {
>                 <td>{{ item[column.field] }}</td>
>               }
>             }
>           }```

This doesn't work for me. Only the first condition is evaluated. So in my case:

```
@case (QuestionType.FileUpload || QuestionType.MultipleFileUpload) {
    <app-case-summary-question-file-upload [attachments]="question.attachments" />
}
```

`QuestionType.MultipleFileUpload` is ignored. If I make both cases explicit it works as expected:

```
@case (QuestionType.FileUpload) {
    <app-case-summary-question-file-upload [attachments]="question.attachments" />
}
@case (QuestionType.MultipleFileUpload) {
    <app-case-summary-question-file-upload [attachments]="question.attachments" />
}
```

**Comment by joafic** [comment]
Still no solution for this case? It would be nice to have something like this:
@case(1) {
//code
}
@case (2), @case (3) {
 //other code
}

**Comment by thaua-engelmann** [comment]
No solution to this day. Only possible workaround is using @if and duplicating the content.

**Comment by arndwestermann** [comment]
I also have such a use case, sad that this still isn't implemented.
Kudos to this fella [here](https://stackoverflow.com/a/78166716/15181911), who has a nice workaround.

**Comment by khadden-compline** [comment]
> I also have such a use case, sad that this still isn't implemented. Kudos to this fella [here](https://stackoverflow.com/a/78166716/15181911), who has a nice workaround.

**Not a fan of the workaround.** 

@petebacondarwin should someone create a new issue for the new "@" template syntax?

**Comment by JeanMeche** [comment]
With `ngSwitch` now deprecated, this is the right issue to track the feature request.

**Comment by steppe87** [comment]
In cases of readability this is a **dearly missed** feature!

**Comment by FrenchFriesFiesta** [comment]
I think this feature should be added urgently, it is standard switch behavior. We just encountered a case where this could save a lot of code duplication.

**Comment by Khumozin** [comment]
We need a solution to this, ASAP.

**Comment by rip222** [comment]
I like this:
`@case (val1, val2, val3) {}`

but this is also fine
```
@case(val1),
@case(val2),
@case(val3) {}
```

doing `@if`s when it should be done with `@switch` is really annoying.

**Comment by eneajaho** [comment]
AI halucinates a lot when it comes to this issue.

**Comment by rip222** [comment]
I hope this is a joke and ai is not the source of feedback for the angular team:)
